src/commands/play.rs

```diff
@@ -77,16 +79,60 @@ pub async fn play(
     summon(ctx, interaction, false).await?;
     let call = manager.get(guild_id).unwrap();
 
-    let query_type = if url.contains("spotify.com") {
-        let spotify = SPOTIFY.lock().await;
-        let spotify = verify(spotify.as_ref(), ParrotError::Other(SPOTIFY_AUTH_FAILED))?;
+    // determine whether this is a link or a query string
+    let query_type = match Url::parse(url) {
+        Ok(url_data) => match url_data.host_str() {
+            Some("open.spotify.com") => {
+                let spotify = SPOTIFY.lock().await;
+                let spotify = verify(spotify.as_ref(), ParrotError::Other(SPOTIFY_AUTH_FAILED))?;
+                Some(Spotify::extract(spotify, url).await?)
+            }
+            Some(other) => {
+                let mut data = ctx.data.write().await;
+                let settings = data.get_mut::<GuildSettingsMap>().unwrap();
+                let guild_settings = settings.entry(guild_id).or_default();
+
+                let other = other.to_string().replace("www.", "");
+
+                if guild_settings.banned_domains.contains(&other)
```

Comment:
Domain parsing will have to be a little more complex than this unfortunately. Here the domain `play.dev` being in the blacklist will mean that `play.dev.com` also gets blacklisted, but these are two different domains. See the following for my best take at a domain parsing algorithm in rust:
```rs
use std::iter;
use url::Url;

fn compare_domains(one: &str, two: &str) -> bool {
    let iterator_one = one.rsplit_terminator(".");
    let iterator_two = two.rsplit_terminator(".");
    let mut matching = true;
    for (element_one, element_two) in iter::zip(iterator_one, iterator_two) {
        println!("{} == {}", element_one, element_two);
        matching = matching && element_one == element_two ;
    }
    matching
}

fn main() {
    let test = Url::parse("http://www.youtube.com./").unwrap();
    let test1 = Url::parse("http://youtube.com/").unwrap();
    println!("{}", compare_domains(test1.domain().unwrap(), test.domain().unwrap()));
}
```

That rsplit_terminator is necessary because `google.com.` is just as valid as `google.com` and will be resolved the same.

Reply:
Wait, I'm dumb. You are using a HashSet for this. That bypasses my issue with the caveat that the user will have to individually add all necessary subdomains to a block/allow list.

Reply:
>  google.com. is just as valid as google.com and will be resolved the same

This issue, however, is still present.

Reply:
Thank you for the review, @StaticRocket. 

I think you're correct, regardless! I do use a HashSet but blocking `play.dev` won't block `play.dev.com`, it has to strictly match. The set just prevents duplicate entries on the form.

Regarding your parsing algorithm, I like it! However, one of the URLs (the one inputted by the user) cannot be parsed via `Url::parse` because it's missing the `http://` portion - panicking with:

```rust
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RelativeUrlWithoutBase', src/main.rs:18:43
```

My question is, should the user manually add all subdomains or should it be done automatically? Also, should the user inputted domains be checked for validity?

Reply:
> Regarding your parsing algorithm, I like it! However, one of the URLs (the one inputted by the user) cannot be parsed via Url::parse because it's missing the http:// portion

That's why it's written to compare string versions of the domain (we shouldn't have to parse the user input as we are using lazy matching). We also allow the user to wildcard match all domains by using the domain entry `.` with that system.

> My question is, should the user manually add all subdomains or should it be done automatically? Also, should the user inputted domains be checked for validity?

I believe subdomain's should inherit parent domain status unless explicitly stated otherwise. The "explicitly stated otherwise" part is the only thing my algorithm can't handle because it would have to rank entries based on specificity (g.z.t.example.com should be interpreted after example.com and override example.com if matched).

Reply:
I wonder if there's any nice domain parsing logic we could lift from ftl / pihole ...

Reply:
> That's why it's written to compare string versions of the domain ; ) (we shouldn't have to parse the user input as we are using lazy matching). We also allow the user to wildcard match all domains by using the domain entry . with that system.

I see what you mean, I was misinterpreting your code, sorry!

Piggybacking on your suggestion, I've been testing the PR with the check below, `domain` being the blocked domain and `subdomain` being the request:

```rust
pub fn compare_domains(domain: &str, subdomain: &str) -> bool {
    subdomain == domain || subdomain.ends_with(domain)
}
```

For instance, if `m.youtube.com` is blocked but `youtube.com` is requested, it's allowed. Conversingly, if `youtube.com` is blocked but `m.youtube.com` is requested, it's blocked. If both happen to be blocked, it's not conflicting, even if redundant.

I reckon it should be enough for now? Did you have any other edge case in mind?